src/v/cloud_topics/level_one/compaction/committer.cc

```diff
@@ -10,101 +10,471 @@
+  metastore::compaction_map_t compact_map) {
+    return l1::retry_metastore_op_with_default_rtc(
+      [this, compact_map = std::move(compact_map)]() {
+          return _metastore->compact_objects(
+            *_metadata_builder, std::move(compact_map));
+      },
+      _as);
+}
+
+ss::future<>
+compaction_committer::compaction_job::compact_objects_without_update() {
+    auto replace_res = co_await do_compact_objects(
+      metastore::compaction_map_t{});
+    if (replace_res.has_value()) {
+        vlog(
+          compaction_log.info,
+          "Finalized compaction of tidp {}, job id {} without a compaction "
+          "metadata update",
+          _tp,
+          _id);
+    } else {
+        vlog(
+          compaction_log.warn,
+          "Could not commit object update during compaction of tidp {}, job id "
+          "{}: {}.",
+          _tp,
+          _id,
+          replace_res.error());
+    }
+}
+
+ss::future<> compaction_committer::compaction_job::compact_objects_with_update(
+  chunked_vector<metastore::compaction_update::cleaned_range>
+    new_cleaned_ranges,
+  offset_interval_set removed_tombstone_ranges) {
+    auto compaction_update = metastore::compaction_update{
+      .new_cleaned_ranges = std::move(new_cleaned_ranges),
+      .removed_tombstones_ranges = std::move(removed_tombstone_ranges),
+      .cleaned_at = model::timestamp::now()};
+
+    auto compaction_update_str = fmt::format("{}", compaction_update);
+    metastore::compaction_map_t compact_map;
+    compact_map.emplace(_tp, std::move(compaction_update));
+    auto commit_res = co_await do_compact_objects(std::move(compact_map));
+
+    if (commit_res.has_value()) {
+        vlog(
+          compaction_log.info,
+          "Finalized compaction of tidp {}, job id {} with compaction metadata "
+          "update {}",
+          _tp,
+          _id,
+          compaction_update_str);
+    } else {
+        vlog(
+          compaction_log.warn,
+          "Could not commit metadata update {} for compaction of tidp {}, job "
+          "id {}: {}. Retrying object update without metadata.",
+          compaction_update_str,
+          _tp,
+          _id,
+          commit_res.error());
+        // We couldn't commit the metastore update, but we should at least try
+        // to replace the objects so as not to discard our hard IO work.
+        co_return co_await compact_objects_without_update();
+    }
+}
+
+ss::future<> compaction_committer::compaction_job::do_finalize(
+  chunked_vector<metastore::compaction_update::cleaned_range>
+    new_cleaned_ranges,
+  offset_interval_set removed_tombstone_ranges) {
+    vlog(
+      compaction_log.debug,
+      "Marking job {} for tidp {} as finalized",
+      _id,
+      _tp);
+    _state = compaction_job::state::finalized;
+    _upload_sem.signal();
+
+    vlog(
+      compaction_log.debug,
+      "Awaiting completion of uploads for job {}, tidp {}",
+      _id,
+      _tp);
+
```

Comment:
Is there a guarantee that _staging_file_and_md_infos is non-empty here? Do we need to upload_some() before this?